arangod/Replication2/ReplicatedLog/types.h

```diff
@@ -73,18 +75,31 @@ auto to_string(FollowerState const&) -> std::string_view;
 struct AppendEntriesRequest;
 struct AppendEntriesResult;
 
-enum class AppendEntriesErrorReason {
-  NONE,
-  INVALID_LEADER_ID,
-  LOST_LOG_CORE,
-  MESSAGE_OUTDATED,
-  WRONG_TERM,
-  NO_PREV_LOG_MATCH,
-  PERSISTENCE_FAILURE,
-  COMMUNICATION_ERROR,
+struct AppendEntriesErrorReason {
+  enum class ErrorType {
+    kNone,
+    kInvalidLeaderId,
+    kLostLogCore,
+    kMessageOutdated,
+    kWrongTerm,
+    kNoPrevLogMatch,
+    kPersistenceFailure,
+    kCommunicationError,
+  };
+
+  ErrorType error = ErrorType::kNone;
+  std::optional<std::string> details = std::nullopt;
+
+  [[nodiscard]] auto getDetails() const noexcept -> std::string_view;
+  [[nodiscard]] auto getErrorMessage() const noexcept -> std::string_view;
+  void toVelocyPack(velocypack::Builder& builder) const;
+  [[nodiscard]] static auto fromVelocyPack(velocypack::Slice slice) -> AppendEntriesErrorReason;
+  static auto errorTypeFromString(std::string_view str) -> ErrorType;
+
+  friend auto operator==(AppendEntriesErrorReason const& left, AppendEntriesErrorReason const& right) noexcept -> bool = default;
 };
 
-[[nodiscard]] auto to_string(AppendEntriesErrorReason reason) noexcept -> std::string_view;
+[[nodiscard]] auto to_string(AppendEntriesErrorReason const& reason) noexcept -> std::string_view;
```

Comment:
I'd prefer having this function accepting a `AppendEntriesErrorReason::ErrorType` parameter instead of the whole struct. It is overly specific.